### DISCUSSION TITLE
what is the difference between JansPerson, JansCustomPerson and eduPerson

### DISCUSSION THREAD

**Discussion by dhanan-np** [JanssenProject/jans#4750] [Q&A] [answered]
While going through the attributes, I came across with the following attributes I could not differentiate between them and the purpose of their existence separately.
1. eduPerson
2. JansPerson
3. JansCustomPerson

Any reference will be highly appreciated.

**Comment by nynymike** [comment, marked as answer]
If you are using LDAP, you can check the schema ldif files in `/opt/jans/opendj/config/schema` and find the respective objectclasses.

* [eduPerson](https://software.internet2.edu/eduperson/internet2-mace-dir-eduperson-201602.html) is a schema defined by Internet2, used primarily in Higher Ed. 
* [jansPerson](https://docs.jans.io/v1.0.12/admin/reference/database/ldap-schema/#jansperson) 
* "jansCustomPerson" is a vestigial artifact -- it was originally a sample of how you could add a custom objectclass. And then people started using it to store their customer attributes.

**Reply by dhanan-np** [reply]
Okay, 
So if we want to create an end user for our application. Who is primarily going to consume the services (lets say Music App). 
How do we create those kind of users. I believe, we will have to do some configuration by which when the user is created will be of the type **jansPerson** instead of **eduPerson**?
Also, I was unable to understand the reason why uid is **MUL** for **KEY**. What if two person with same uid is registered accidently. Which one will be returned?

**Comment by nynymike** [comment]
`uid` should be deterministic--no duplicates. Uniqueness rules are defined in the respective database. 

If your Music App has a lot of specific claims, you should create a new objectcalss, such as `musicAppPerson`  and register the additional attributes against it. Then add it as an additional objectclass.

**Reply by dhanan-np** [reply]
that's great thing to know. How do I create custom objectClass? I believe it must be extending jansPerson to inherit primary attributes.

My current system is running on **MySql**

**Comment by yurem** [comment]
In case if DB is `SQL/Couchbase/Spanner` you can add additional attributes to `jansPerson` table and use them. This will be similar to LDAP where DB stores all user attributes in one entry. Additional attributes not affects server functionality.

**Reply by dhanan-np** [reply]
I added the attribute and tried to update a user with the attribute but got this error-

Caused by: io.jans.orm.exception.operation.PersistenceException: Failed to find attribute type for '**sampleService3**'
        at io.jans.orm.sql.operation.impl.SqlOperationServiceImpl.updateEntryImpl(**SqlOperationServiceImpl.java:243**) ~[jans-orm-sql-1.0.12.jar:?]
        at io.jans.orm.sql.operation.impl.SqlOperationServiceImpl.updateEntry(**SqlOperationServiceImpl.java:222**) ~[jans-orm-sql-1.0.12.jar:?]
        at io.jans.orm.sql.impl.SqlEntryManager.merge(SqlEntryManager.java:269) ~[jans-orm-sql-1.0.12.jar:?]

**Reply by dhanan-np** [reply]
one more question on this. Can we use this attribute in claims?

**Reply by dhanan-np** [reply]
I tried one more way...
Added the attribute **sampleService1** using **TUI** and then added that as a column in jansPerson table,
but same error was thrown as previously: 
```
...
Failed to find attribute type for 'sampleService1'
at io.jans.orm.sql.operation.impl.SqlOperationServiceImpl.updateEntryImpl(SqlOperationServiceImpl.java:243)
...
```

**Reply by yurem** [reply]
Did you restart services after DB schema modification? ORM loads schema only one time at startup

**Reply by dhanan-np** [reply]
thanks a lot @yurem its working now after restarting the server.
The values are coming into **urn:ietf:params:scim:schemas:extension:gluu:2.0:User**. I wish I could get them as a normal attribute like others and could mention in the api-parameter **attributes** to see that data only.

@ossdhaval I have noticed one issue with **TUI**. If I have an custom attribute with boolean[mysql: tinyint(1)] data-type, it doesn't show as selected when their value is True(1) in the database.

![TUI](https://user-images.githubusercontent.com/131341377/235904821-d563d63c-fb23-45cc-b7e6-aaf869416aba.jpg)

**Reply by ossdhaval** [reply]
@dhanan-np We will check on the TUI issue.

**Reply by ossdhaval** [reply]
Above mentioned TUI behavior is an issue. Confirmed by @devrimyatar . We will track using #4796 

Thanks for highlighting this :handshake:

**Reply by dhanan-np** [reply]
With your permission, I would suggest one thing that, for mysql, there should be something at configuration level because **not everything** that is defined by **tinyint** will be **boolean**. Somewhere it could be expected as int value from 1 to 255. In some of our application we have status defined as tinyint, because some object can have 10 different status in their lifecycle. 
And since we need to restart our server post introduction of any custom attribute, we can apply that change in ORM.
